presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizationManager.java

```diff
@@ -0,0 +1,314 @@
+        this.shardOrganizerDao = onDemandDao(dbi, ShardOrganizerDao.class);
+
+        this.organizer = requireNonNull(organizer, "organizer is null");
+        this.shardManager = requireNonNull(shardManager, "shardManager is null");
+        this.currentNodeIdentifier = requireNonNull(currentNodeIdentifier, "currentNodeIdentifier is null");
+
+        this.enabled = enabled;
+
+        requireNonNull(organizationInterval, "organizationInterval is null");
+        this.organizationIntervalSeconds = max(1, organizationInterval.roundTo(SECONDS));
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (!enabled) {
+            return;
+        }
+
+        if (!started.getAndSet(true)) {
+            startDiscovery();
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        shutdown.set(true);
+        discoveryService.shutdownNow();
+    }
+
+    private void startDiscovery()
+    {
+        discoveryService.scheduleWithFixedDelay(() -> {
+            if (shutdown.get()) {
+                return;
+            }
+
+            try {
+                // jitter to avoid overloading database
+                SECONDS.sleep(ThreadLocalRandom.current().nextLong(1, 5 * 60));
+
+                log.info("Running shard organizer...");
+                Set<Long> tableIds = discoverAndInitializeTablesToOrganize();
+                submitJobs(tableIds);
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            catch (Throwable t) {
+                log.error(t, "Error running shard organizer");
+            }
+        }, 0, 5, MINUTES);
+    }
+
+    @VisibleForTesting
+    Set<Long> discoverAndInitializeTablesToOrganize()
+    {
+        Set<Long> enabledTableIds = metadataDao.getOrganizationEligibleTables();
+
+        Set<TableOrganizationInfo> tableOrganizationInfo = shardOrganizerDao.getNodeTableOrganizationInfo(currentNodeIdentifier);
+        Map<Long, TableOrganizationInfo> organizationInfos = Maps.uniqueIndex(tableOrganizationInfo, TableOrganizationInfo::getTableId);
+
+        // If this is the first time organizing a table, initialize the organization info for it
+        Set<Long> newTableIds = difference(enabledTableIds, organizationInfos.keySet());
+        newTableIds.forEach(tableId -> shardOrganizerDao.insertNode(currentNodeIdentifier, tableId));
+
+        ImmutableSet.Builder<Long> tableIds = ImmutableSet.builder();
+        for (Long tableId : enabledTableIds) {
+            TableOrganizationInfo info = organizationInfos.get(tableId);
+            if (info == null || shouldRunOrganization(tableId, info)) {
+                tableIds.add(tableId);
+            }
+        }
+        return tableIds.build();
+    }
+
+    private void submitJobs(Set<Long> tableIds)
+    {
+        for (long tableId : tableIds) {
+            shardOrganizerDao.updateLastStartTime(currentNodeIdentifier, tableId);
```

Comment:
This should run at the end. Otherwise, if the job fails (or the worker is restarted), it won't run for the full interval (which could be very long)